modules/renter/download.go

```diff
@@ -1,805 +1,505 @@
-		// saturated with fewer workers, then the more expensive ones should be
-		// eliminated.
-
-		ds.availableWorkers = append(ds.availableWorkers, worker)
 	}
-	r.mu.Unlock(id)
-
-	// Add new chunks to the extent that resources allow.
-	r.managedScheduleNewChunks(ds)
-
-	// Check for incomplete chunks, and assign workers to them where possible.
-	r.managedScheduleIncompleteChunks(ds)
-
-	// Wait for workers to return after downloading pieces.
-	r.managedWaitOnDownloadWork(ds)
-}
-
-// managedScheduleIncompleteChunks iterates through all of the incomplete
-// chunks and finds workers to complete the chunks.
-// managedScheduleIncompleteChunks also checks whether a chunk is unable to be
-// completed.
-func (r *Renter) managedScheduleIncompleteChunks(ds *downloadState) {
-	var newIncompleteChunks []*chunkDownload
-loop:
-	for _, incompleteChunk := range ds.incompleteChunks {
-		// Drop this chunk if the file download has failed in any way.
-		incompleteChunk.download.mu.Lock()
-		downloadComplete := incompleteChunk.download.downloadComplete
-		incompleteChunk.download.mu.Unlock()
-		if downloadComplete {
-			// The download has most likely failed. No need to complete this
-			// chunk.
-			ds.activePieces--                                       // For the current incomplete chunk.
-			ds.activePieces -= len(incompleteChunk.completedPieces) // For all completed pieces.
-
-			// Clear the set of completed pieces so that we do not
-			// over-subtract if the above code is run multiple times.
-			incompleteChunk.completedPieces = make(map[uint64][]byte)
-			continue
-		}
-
-		// Try to find a worker that is able to pick up the slack on the
-		// incomplete download from the set of available workers.
-		for i, worker := range ds.availableWorkers {
-			scheduled, exists := incompleteChunk.workerAttempts[worker.contract.ID]
-			if scheduled || !exists {
-				// Either this worker does not contain a piece of this chunk,
-				// or this worker has already been scheduled to download a
-				// piece for this chunk.
-				continue
-			}
-
-			piece, exists := incompleteChunk.download.pieceSet[incompleteChunk.index][worker.contract.ID]
-			if !exists {
-				continue
-			}
+	params.file.mu.Unlock()
 
-			dw := downloadWork{
-				dataRoot:      piece.MerkleRoot,
-				pieceIndex:    piece.Piece,
-				chunkDownload: incompleteChunk,
-				resultChan:    ds.resultChan,
-			}
-			incompleteChunk.workerAttempts[worker.contract.ID] = true
-			ds.availableWorkers = append(ds.availableWorkers[:i], ds.availableWorkers[i+1:]...)
-			ds.activeWorkers[worker.contract.ID] = struct{}{}
-			select {
-			case worker.priorityDownloadChan <- dw:
-			default:
-				r.log.Critical("Download work not immediately received by worker")
-			}
-			continue loop
-		}
+	// Queue the downloads for each chunk.
+	writeOffset := int64(0) // where to write a chunk within the download destination.
+	for i := minChunk; i <= maxChunk; i++ {
+		d.chunksRemaining++
```

Comment:
If we change the loop such that chunks can fail, or something to that effect, it may impact how this variable is used. I think better to increase the counter as you add chunks.

Reply:
oh actually, this could cause errors. If the download only has 2 chunks and those 2 chunks finish before it queues the third, it's going to think the download is complete. hmm. Guess it does need to be outside the loop.